include/deal.II/base/quadrature_point_data.h

```diff
@@ -0,0 +1,808 @@
+   *
+   * @note Subsequent calls of this function with the same @p cell will not
+   * alter the objects associated with it. In order to remove the stored data,
+   * use the erase() function.
+   *
+   * @note It is possible to use different types @p T for different cells which
+   * may reflect, for example, different constitutive models of continuum mechanics
+   * in different parts of the domain.
+   */
+  template<typename T=DataType>
+  void initialize(const CellIteratorType &cell,
+                  const unsigned int number_of_data_points);
+
+  /**
+   * Same as above but for a range of iterators starting at @p cell_start
+   * until, but not including, @p cell_end for all locally owned cells, i.e.
+   * for which `is_locally_owned()==true`
+   */
+  template<typename T=DataType>
+  void initialize(const CellIteratorType &cell_start,
+                  const CellIteratorType &cell_end,
+                  const unsigned int number);
+
+  /**
+   * Removes data stored at the @p cell. Returns true if the data was removed.
+   * If no data is attached to the @p cell, this function will not do anything
+   * and returns false.
+   *
+   * @note In debug mode, this function will also check that the are no
+   * outstanding references to the data stored on this cell. That is to say,
+   * that the only references to the stored data are that made by this class.
+   */
+  bool erase(const CellIteratorType &cell);
+
+  /**
+   * Clear all the data stored in this object.
+   */
+  void clear();
+
+  /**
+   * Get a vector of the data located at @p cell .
+   * A possible additional typename @p T is the class to which the base class
+   * DataType could be cast. Since @p DataType is stored as shared pointers,
+   * there is minimal overhead in returning a vector by value instead of by reference.
+   * This allows flexibility if class @p T is not the same as @p DataType on a
+   * cell-by-cell basis.
+   */
+  template<typename T=DataType>
+  CellVector<T> get_data(const CellIteratorType &cell);
+
+  /**
+   * Get a vector of constant pointers to data located at @p cell .
+   * A possible additional typename @p T is the class to which the base class
+   * DataType could be cast. Since @p DataType is stored as shared pointers,
+   * there is minimal overhead in returning a vector by value instead of by reference.
+   * This allows flexibility if class @p T is not the same as @p DataType on a
+   * cell-by-cell basis.
+   */
+  template<typename T=DataType>
+  CellVector<const T> get_data(const CellIteratorType &cell) const;
+
+private:
+  /**
+   * A map to store a vector of data on a cell.
+   */
+  std::map<CellIteratorType, CellVector<DataType> > map;
+
+};
+
+
+/**
+ * An abstract class which specifies requirements for data on
+ * a single quadrature point to be transferable during refinement.
+ * One possible implementation is via the restriction/prolongation matrices
+ * as implemented in ContinuousQuadratureDataTransfer class.
```

Comment:
Can you give a high level overview of what this class does here? Maybe something along the lines of
```
  This class provides a framework by which derived classes representing data at quadrature points can declare how many scalar values they store, and then implement functions that pack and unpack these scalars into arrays that can be transferred from one cell to another, or one processor to another upon mesh refinement and repartitioning.
```